core/network/impl/state_protocol_observer_impl.cpp

```diff
@@ -0,0 +1,159 @@
+    unsigned size = 0;
+    KeyValueStateEntry entry;
+    // main state storage hash is marked with zeros in response
+    entry.state_root =
+        storage::trie::RootHash::fromSpan(std::vector<uint8_t>(32, 0)).value();
+
+    StateResponse response;
+    response.entries.emplace_back(std::move(entry));
+
+    // First key would contain main state storage key (child state storage hash)
+    // Second key is child state storage key
+    if (request.start.size() == 2) {
+      const auto &parent_key = request.start[0];
+      const auto &child_prefix = storage::kChildStorageDefaultPrefix;
+      if (parent_key.size() < child_prefix.size()
+          || parent_key.subbuffer(0, child_prefix.size()) != child_prefix) {
+        return Error::INVALID_CHILD_ROOTHASH;
+      }
+      if (auto value_res = batch->tryGet(parent_key);
+          value_res.has_value() && value_res.value().has_value()) {
+        auto hash =
+            storage::trie::RootHash::fromSpan(value_res.value().value().get())
+                .value();
+        OUTCOME_TRY(
+            entry_res,
+            this->getEntry(hash, request.start[1], MAX_RESPONSE_BYTES - size));
+        response.entries.emplace_back(std::move(entry_res.first));
+        size += entry_res.second;
+      } else {
+        return Error::NOTFOUND_CHILD_ROOTHASH;
+      }
+    }
+
+    if (res.has_value()) {
+      const auto &child_prefix = storage::kChildStorageDefaultPrefix;
+      while (cursor->key().has_value() && size < MAX_RESPONSE_BYTES) {
+        if (auto value_res = batch->tryGet(cursor->key().value());
+            value_res.has_value()) {
+          const auto &value = value_res.value();
+          auto &entry = response.entries.front();
+          entry.entries.emplace_back(
+              StateEntry{cursor->key().value(), value.value()});
+          size += entry.entries.back().key.size()
+                  + entry.entries.back().value.size();
+          // if key is child state storage hash iterate child storage keys
+          if (cursor->key().value().size() > child_prefix.size()
+              && cursor->key().value().subbuffer(0, child_prefix.size())
+                     == child_prefix) {
+            OUTCOME_TRY(hash,
+                        storage::trie::RootHash::fromSpan(
+                            value_res.value().value().get()));
+            OUTCOME_TRY(entry_res,
+                        this->getEntry(
+                            hash, common::Buffer(), MAX_RESPONSE_BYTES - size));
+            response.entries.emplace_back(std::move(entry_res.first));
+            size += entry_res.second;
+            // not complete means response bytes limit exceeded
+            // finish response formation
+            if (not entry_res.first.complete) {
+              break;
+            }
+          }
+        }
+        res = cursor->next();
+      }
+      response.entries.front().complete = not cursor->key().has_value();
```

Comment:
Do you sure it should be front()? Should not be back()?

Reply:
Nope. First entry is always keys from main batch, others are for child storages.